File: frontend/lib/settlement.ts
/**
 * The settlement engine — prototype mirror of the C# domain engine.
 *
 * PROTOTYPE ONLY. Phase 2 deletes this file: the real engine lives in
 * Owem.Domain and this app calls POST /events/{id}/settlement instead.
 * It exists so the clickable prototype shows real, consistent numbers that
 * respond to what a tester actually taps, rather than hardcoded amounts.
 *
 * It keeps the two properties the real engine must have:
 *   INVARIANT 1 — it reads only USER_CONFIRMED and SYSTEM_COMPUTED values.
 *                 Given anything AI_SUGGESTED it throws.
 *   The parts sum exactly to the total, asserted before it returns.
 */
import { allocate, assertSumsTo, cents, sum, ZERO } from './money.ts';
import type {
  Cents,
  ItemAssignment,
  Participant,
  Receipt,
  ReceiptItem,
  SettlementLine,
} from './types.ts';

export const ENGINE_VERSION = 'settlement-prototype-1.0.0';

export class UnconfirmedInputError extends Error {
  readonly itemIds: string[];
  constructor(itemIds: string[]) {
    super(`${itemIds.length} value(s) are still AI_SUGGESTED. Confirm them first.`);
    this.name = 'UnconfirmedInputError';
    this.itemIds = itemIds;
  }
}

export class UnassignedItemsError extends Error {
  readonly itemIds: string[];
  constructor(itemIds: string[]) {
    super(`${itemIds.length} item(s) are not assigned to anyone.`);
    this.name = 'UnassignedItemsError';
    this.itemIds = itemIds;
  }
}

export type SettlementInput = {
  receipt: Receipt;
  items: ReceiptItem[];
  assignments: ItemAssignment[];
  participants: Participant[];
};

export function computeSettlement(input: SettlementInput): SettlementLine[] {
  const { receipt, items, assignments, participants } = input;

  // INVARIANT 1: the guard clause at the engine boundary.
  const unconfirmed = items.filter((i) => i.provenance === 'AI_SUGGESTED').map((i) => i.id);
  if (unconfirmed.length > 0) throw new UnconfirmedInputError(unconfirmed);
  if (receipt.taxProvenance === 'AI_SUGGESTED') throw new UnconfirmedInputError(['tax']);

  const unassigned = items.filter((i) => !assignments.some((a) => a.itemId === i.id));
  if (unassigned.length > 0) throw new UnassignedItemsError(unassigned.map((i) => i.id));

  const ids = participants.map((p) => p.id);
  const index = new Map(ids.map((id, i) => [id, i]));
  const itemsSubtotal = ids.map(() => 0);

  // 1. Every item's price is split across the people on it, by weight.
  for (const item of items) {
    const on = assignments.filter((a) => a.itemId === item.id);
    const shares = allocate(item.totalPrice, on.map((a) => a.weight));
    assertSumsTo(shares, item.totalPrice, `item ${item.normalizedName}`);
    on.forEach((a, i) => {
      const at = index.get(a.participantId);
      if (at === undefined) throw new Error(`assignment references unknown participant`);
      itemsSubtotal[at] += shares[i];
    });
  }

  const subtotal = cents(itemsSubtotal.reduce((a, b) => a + b, 0));

  // 2. Tax, tip and discount follow the policy.
  //    PROPORTIONAL weights by what each person ordered; EQUAL splits per head.
  const weights =
    receipt.tipPolicy === 'EQUAL' ? ids.map(() => 1) : itemsSubtotal.map((v) => v);
  const safeWeights = weights.every((w) => w === 0) ? ids.map(() => 1) : weights;

  const taxShares = allocate(receipt.tax, safeWeights);
  const tipShares = allocate(receipt.tip, safeWeights);
  const discountShares = allocate(receipt.discount, safeWeights);

  assertSumsTo(taxShares, receipt.tax, 'tax');
  assertSumsTo(tipShares, receipt.tip, 'tip');
  assertSumsTo(discountShares, receipt.discount, 'discount');

  const lines: SettlementLine[] = ids.map((participantId, i) => ({
    participantId,
    itemsSubtotal: cents(itemsSubtotal[i]),
    taxShare: taxShares[i],
    tipShare: tipShares[i],
    discountShare: discountShares[i],
    amountOwed: cents(itemsSubtotal[i] + taxShares[i] + tipShares[i] - discountShares[i]),
  }));

  // 3. The whole point: the parts sum to the whole, exactly.
  const expected = cents(subtotal + receipt.tax + receipt.tip - receipt.discount);
  assertSumsTo(lines.map((l) => l.amountOwed), expected, 'settlement');

  return lines;
}

/** What the receipt says the bill came to, from its own parts. */
export function receiptTotalFromParts(items: ReceiptItem[], receipt: Receipt): Cents {
  return cents(sum(items.map((i) => i.totalPrice)) + receipt.tax + receipt.tip - receipt.discount);
}

/** The redundancy a receipt gives you for free — use it as a checksum. */
export function checkReceipt(items: ReceiptItem[], receipt: Receipt) {
  const itemsTotal = sum(items.map((i) => i.totalPrice));
  const computed = receiptTotalFromParts(items, receipt);
  return {
    itemsTotal,
    computed,
    printed: receipt.total,
    balances: computed === receipt.total,
    drift: cents(computed - receipt.total),
  };
}

/** Money still out, per participant, after payments. Never below zero. */
export function outstandingFor(
  line: SettlementLine | undefined,
  paid: Cents,
): Cents {
  if (!line) return ZERO;
  return cents(Math.max(0, line.amountOwed - paid));
}

File: backend/owem/settlement.py
from decimal import Decimal
from uuid import UUID

from owem.errors import unprocessable
from owem.models import Assignment, Receipt, ReceiptItem, SettlementLine
from owem.money import allocate, assert_sums_to, to_cents, to_money

ENGINE_VERSION = "settlement-1.0.0"

CONFIDENCE_FLOOR = Decimal("0.85")
RECEIPT_CONFIDENCE_FLOOR = Decimal("0.90")
TOLERANCE = Decimal("0.05")
MAX_TAX_RATE = Decimal("0.20")


def compute_settlement(
    receipt: Receipt,
    items: list[ReceiptItem],
    assignments: list[Assignment],
    participant_ids: list[UUID],
) -> list[SettlementLine]:
    reject_unconfirmed(receipt, items)
    reject_unassigned(items, assignments)

    people = list(dict.fromkeys(participant_ids))
    position = {person: i for i, person in enumerate(people)}
    subtotals = [0] * len(people)

    for item in items:
        on_item = [a for a in assignments if a.item_id == item.id]
        for a in on_item:
            if a.participant_id not in position:
                raise unprocessable(
                    "UNKNOWN_PARTICIPANT",
                    f"Assignment names somebody not on this event: {a.participant_id}",
                )
        shares = allocate(to_cents(item.total_price), [a.weight for a in on_item])
        for a, share in zip(on_item, shares, strict=True):
            subtotals[position[a.participant_id]] += share

    weights = charge_weights(receipt.tip_policy, subtotals)
    tax = allocate(to_cents(receipt.tax), weights)
    tip = allocate(to_cents(receipt.tip), weights)
    discount = allocate(to_cents(receipt.discount), weights)

    lines = [
        SettlementLine(
            participant_id=person,
            items_subtotal=to_money(subtotals[i]),
            tax_share=to_money(tax[i]),
            tip_share=to_money(tip[i]),
            discount_share=to_money(discount[i]),
            amount_owed=to_money(subtotals[i] + tax[i] + tip[i] - discount[i]),
        )
        for i, person in enumerate(people)
    ]

    expected = (
        sum(subtotals) + to_cents(receipt.tax) + to_cents(receipt.tip) - to_cents(receipt.discount)
    )
    assert_sums_to([to_cents(line.amount_owed) for line in lines], expected, "settlement")
    return lines


def reject_unconfirmed(receipt: Receipt, items: list[ReceiptItem]) -> None:
    unconfirmed = [str(item.id) for item in items if item.provenance == "AI_SUGGESTED"]
    fields = [] if receipt.tax_provenance != "AI_SUGGESTED" else ["tax"]
    if unconfirmed or fields:
        raise unprocessable(
            "UNCONFIRMED_INPUT",
            f"{len(unconfirmed) + len(fields)} value(s) are still AI_SUGGESTED. "
            "Confirm them first.",
            itemIds=unconfirmed,
            fields=fields,
        )


def reject_unassigned(items: list[ReceiptItem], assignments: list[Assignment]) -> None:
    assigned = {a.item_id for a in assignments}
    missing = [str(item.id) for item in items if item.id not in assigned]
    if missing:
        raise unprocessable(
            "UNASSIGNED_ITEMS",
            f"{len(missing)} item(s) are not assigned to anyone.",
            itemIds=missing,
        )


def charge_weights(policy: str, subtotals: list[int]) -> list[Decimal]:
    if policy == "EQUAL" or not any(subtotals):
        return [Decimal(1)] * len(subtotals)
    return [Decimal(value) for value in subtotals]


def settlement_total(lines: list[SettlementLine]) -> Decimal:
    return sum((line.amount_owed for line in lines), Decimal("0.00"))


def check_receipt(items: list[ReceiptItem], receipt: Receipt) -> list[str]:
    problems = []
    line_total = sum((item.total_price for item in items), Decimal("0"))
    computed = line_total + receipt.tax + receipt.tip - receipt.discount

    if abs(computed - receipt.total) > TOLERANCE:
        problems.append(
            f"The lines and charges come to {computed} but the receipt says {receipt.total}."
        )
    if line_total > 0 and receipt.tax > line_total * MAX_TAX_RATE:
        problems.append(
            f"Tax of {receipt.tax} is more than 20% of {line_total} - that is a misread, not a tax."
        )
    return problems
